lib/krikri/ldp/invalidatable.rb

```diff
@@ -0,0 +1,102 @@
+module Krikri::LDP
+  ##
+  # Implements invalidation for `Krikri::LDP::Resource`s. This is different 
+  # from deletion, in that the resource continues to respond `200 OK`, and 
+  # return the representation, Nothing is removed from the LDP server.
+  # 
+  # Works as a mixin to `Krikri::LDP::Resource`, assuming an implementation of
+  # `#rdf_source`, which may simply return `self`.
+  #
+  # @example invalidating a resource
+  #   class MyResource
+  #     include Krikri::LDP::Resource
+  #     include Krikri::LDP::Invalidatable
+  #
+  #     def rdf_subject
+  #       @rdf_subject ||= RDF::URI('http://example.com/ldp/a/resource/path')
+  #     end
+  #   end
+  #  
+  #   invalidatable_resource = MyResource.new
+  #   # the resource must exist before it can be invalidated! 
+  #   invalidatable_resource.save
+  #
+  #   invalidatable_resource.invalidate!
+  #   invalidatable_resource.invalidated? # => true
+  #   invalidatable_resource.invalidated_at_time 
+  #   # => Thu, 03 Dec 2015 10:27:45 -0800
+  #
+  # @see http://www.w3.org/TR/2013/REC-prov-dm-20130430/#term-Invalidation 
+  #   for documentation on PROV invalidation
+  module Invalidatable
+    # @see RDF::PROV
+    INVALIDATED_BY_URI = RDF::PROV.wasInvalidatedBy
+    INVALIDATED_TIME_URI = RDF::PROV.invalidatedAtTime
+
+    ##
+    # Invalidates the resource by marking it with a `prov:invalidatedAtTime`. If
+    # an `RDF::Term` is passed as the first argument, that term is used as the 
+    # value of `prov:wasInvalidatedBy`.
+    #
+    # @example invalidating with an activity
+    #   invalidatable_resource.invalidate!(RDF::URI('http://example.org/moomin'))
+    #   invalidatable_resource.was_invalidated_by
+    #   # => #<RDF::URI:0x2acab846109c URI:http://example.org/moomin>
+    #
+    # @param activity_uri [RDF::Term] a URI for the invalidating activity. If 
+    #   none is given, this defaults to `nil` and no `prov:wasInvalidatedBy`
+    #   statement is added.
+    #
+    # @raise [RuntimeError] when the resource does not exist or is already 
+    #   invalid
+    # @return [void]
+    def invalidate!(activity_uri = nil)
+      raise "Cannot invalidate #{rdf_subject}, does not exist." unless exists?
+
+      # force a reload unless we have cached an invalidatedAtTime
+      rdf_source.get({}, true) unless invalidated?
+      # we check invalidated again in case the reload came back invalid
+      raise "Cannot invalidate #{rdf_subject}, already invalid." if invalidated?
```

Comment:
Can you elaborate on why an exception is raised in this case?

Perhaps without realizing some subtleties to the usage of `!` I'd expect this to be idempotent in the fashion of this example:
```
irb(main):006:0> s = "A"
=> "A"
irb(main):007:0> s.downcase!
=> "a"
irb(main):008:0> s
=> "a"
irb(main):009:0> s = "a"
=> "a"
irb(main):010:0> s.downcase!  # <-- note this case
=> nil
irb(main):011:0> s
=> "a"
```
I'm thinking of how downcasing "a" does not result in an exception.  It returns `nil`, which I'm not sure is intuitive, but it doesn't cause an exception in a routine that might be processing items without wanting to check each one for whether it's already in the desired state. For example, I could see wanting to invalidate a number of records, having that fail partway through due to some random error elsewhere in the application, and then wanting to rerun the operation on the same list.